### PR TITLE
chore(core): improve ruleset typings

### PR DESCRIPTION
**Pull request by P0lip**
Extract typing-related changes from #2026

**Checklist**

- [x] Tests added / updated
- [ ] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/ruleset/types.ts [hunk 1]
```diff
@@ -112,3 +112,10 @@ export type RulesetDefinition = Readonly<
       }
   >
 >;
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type Stringifable<T> = T extends object
+  ? {
+      [P in keyof T]: Stringifable<T[P]> | { toJSON?(): Stringifable<T[P]> };
+    }
+  : T;
```
Comment (marbemac) [inline packages/core/src/ruleset/types.ts:121]: Neat

File: tsconfig.build.json [hunk 2]
```diff
@@ -1,7 +1,9 @@
 {
   "include": [],
   "extends": "./tsconfig.json",
-  "composite": true,
+  "compilerOptions": {
+    "composite": true
+  },
```
Comment (marbemac) [inline tsconfig.build.json:6]: 🤦‍♂️

File: packages/core/package.json [hunk 3]
```diff
@@ -40,12 +44,12 @@
     "@stoplight/spectral-parsers": "^1.0.0",
     "@stoplight/spectral-ref-resolver": "^1.0.0",
     "@stoplight/spectral-runtime": "^1.0.0",
-    "@stoplight/types": "12.3.0",
+    "@stoplight/types": "13.0.0",
```
Comment (marbemac) [inline packages/core/package.json:47]: Plans to bump to 13 in the other spectral packages?

**Review by marbemac — APPROVED** [review]

File: packages/core/package.json (hunk 3 above)
Comment (P0lip) [reply packages/core/package.json:47]: Yeah, I intend to do it

File: tsconfig.build.json (hunk 2 above)
Comment (P0lip) [reply tsconfig.build.json:6]: Yeah, luckily newer TS caught this

**Pull request merged**